---
title: "evidence2"
output: html_document
date: "2025-05-02"
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

First, we must load all of the libraries that we will use for this project. Some of these have to be installed through BioCManager and thus we must include the appropriate installation instructions.

```{r echo = F, results = 'hide'}
# BioCManager library installation
#if (!requireNamespace("BiocManager", quietly = TRUE)){
#  install.packages("BiocManager")
#  BiocManager::install("msa")
#  BiocManager::install("DECIPHER")
#  BiocManager::install("phangorn")
#}

# Library loading
library(ade4)
library(ape)
library(adegenet)
library(Biostrings)
library(DECIPHER)
library(ggplot2)
library(reshape2)
library(dplyr)
library(stringr)
library(tidyr)
library(phangorn)
library(ggplot2)
library(ggthemes)
library(ggrepel)
library(msa)

```

Then, for each of the countries we are loading genomes from, we must read their DNA sequence and store them in a vector for further processing.

This whole block is in charge of the following:
* Read all genome sequences from the selected countries
* Convert them to DNA binaries
* Create a data frame containing the solicited characteristics such as base frequency or genome length
* The data frame is used to print the characteristics of each present genome such as base frequency and base length


```{r}
# List of countries where COVID killed the most people
countries <- c("usa", "china", "india", "france", "germany", "brazil", "south_korea", "japan", "italy", "uk")

# Map the names to DNA String Sets
sequences <- sapply(countries, function(x) readDNAStringSet(paste("./sequences/", x, ".fasta", sep="")))

# Convert to DNABin types
dna_bins <- do.call(c, sapply(sequences, as.DNAbin))

# Labeling and Display
country_titles <- sapply(countries, function(string) str_to_title(str_replace(string, "_", " ")))

# Create a  Dataframe from the Sequences
sequence_frame <- data.frame(
  country_key = countries,
  country_title  = country_titles,
  sequence_length = sapply(sequences, width),
  a_frequency = sapply(sequences, function(seq) letterFrequency(seq, 'A')),
  c_frequency = sapply(sequences, function(seq) letterFrequency(seq, 'C')),
  g_frequency = sapply(sequences, function(seq) letterFrequency(seq, 'G')),
  t_frequency = sapply(sequences, function(seq) letterFrequency(seq, 'T'))
)

# Print Genome Sequence Lengths
print("The sequence lengths for each genome are as follows")
sequence_frame[3]

# Print Genome Sequence Nitrogenous Base Frequencies
print("The frequencies for each base in each genome are as follows:")
sequence_frame[c(4, 5, 6, 7)]

```

Then, the data frame is used to generate a frequency graph for each genome in the sequence vector. There is a set of 4 bars for each present genome.

```{r}
# Plot Frequencies
ggplot(
  data = sequence_frame %>% gather(Frequencies, Frequency, -country_title, -country_key, -sequence_length),
  aes(x = country_title, y = Frequency, fill = Frequencies)
) +
  geom_bar(stat = 'identity', position = 'dodge') +
  xlab("Countries") +
  scale_fill_discrete(
    labels = c(
      "a_frequency" = "Number of A Bases",
      "c_frequency" = "Number of C Bases",
      "g_frequency" = "Number of G Bases",
      "t_frequency" = "Number of T Bases"
  ))

```
Now, we read the FASTA files again because we didn't find a way to convert them directly to a DNABin formal class. Using this, we generate a phylogenetic tree for each of the genomes:

* Read the FASTA files into a DNAStringSet containing DNA Binaries
* Load the case and mortality data for each country that we analyzed
* Determine the mortality rate of each virus inper country
* Align the genome sequences using AlignSeqs.
* There are also pre-aligned files in the sequences folder. I aligned these using MAFFT.
* If we couldn't align them, then we stop the program
* Then we create a dna distribution using the TN93 model, this means a distribution that allows us to calculate the divergence bewteen DNA sequences. This model is based on the Tamura-Nei 1993 model.
* Then, we create a phylogenetic tree from the `nj` function. This applies the neighbor-joining algorithm to cluster the most closely related sequences from tge provided distance matrix and makes branch lengths proportional to the evolutionary divergence calculated via TN93.
* We plot the tree, along with bootstrap values which indicate how confident we are in wether or not these two branches are related. The more closely related, the more green the value shows up, else we side towards red.
* Finally, we generate an adjacency matrix from our calculated DNA distribution, which allows us to see, graphically, how closely related two sequences are.
* As an extra, we have added a dot at each tip to indicate the mortality rate of the given virus. This allows us to see how deadly viruses become as they evolve.

```{r}
# Read FASTAs Again but in a Different Format (idealy, this shouldn't be the case but we couldn't find how to do it)
countries <- c("usa", "china", "india", "france", "germany", "brazil", "south_korea", "japan", "italy", "uk")

# Place Sequences into a DNAStringSet
all_sequences <- DNAStringSet()
for (country in countries) {
  file_path <- paste0("./sequences/", country, ".fasta")
  seq <- readDNAStringSet(file_path)
  all_sequences <- c(all_sequences, seq)
}

# Load Mortality Rates
# 3. Prepare mortality data with EXACT MATCH to tip labels
mortality_data <- data.frame(
  label = countries,  # Must match tree$tip.label
  Cases = c(103436829, 99380363, 45043415, 39016278, 38437756, 37511921, 
            34571873, 33803572, 26826486, 24992089),
  Deaths = c(1201488, 122358, 533641, 168091, 174979, 702116, 
             35934, 74694, 197542, 232112)
  
)

mortality_data$Rates = mortality_data$Deaths / mortality_data$Cases

# Align DNA Sequences
aligned <- AlignSeqs(all_sequences)
dna <- as.matrix(as.DNAbin(aligned))

# Stop if Our Sequences are not Aligned
stopifnot(length(unique(ncol(dna))) == 1)

# DNA Distribution
dna_distro <- dist.dna(dna, model = "TN93")

# Create NJ Tree From DNA Distro
phylotree <- nj(dna_distro)

# Calculate Bootstrapping Values
boots <- boot.phylo(
  phylotree, 
  dna, 
  function(e) root(nj(dist.dna(e, model = "TN93")), 1),
  B = 100,
  quiet = TRUE
)

# Plot our Phylo Tree with Bootstrap Values
cpalette <- colorRampPalette(c("red","yellow","green","blue"))
mortalityPalette <- colorRampPalette(c("red", "blue"))
plot(phylotree, cex = 0.6, main = "NJ Tree")
nodelabels(boots, bg = "white", cex = 0.6, col=transp(num2col(boots, col.pal=cpalette),.7))
tiplabels(frame = "none", pch = 20, col = transp(num2col(mortality_data$Rates, col.pal=mortalityPalette)))

# Convert our DNA Distribution and Plot it in a Table
adj_matrix <- as.matrix(dna_distro)
table.paint(
  adj_matrix,
  cleg = 0,
  clabel.row = 0.25,
  clabel.col = 0.25
)
```


*CONCLUSIONS*

As we can see from the gathered data, we can conclude that some of the viruses we have explored don't necessarily relate to each other, however, many of them have a high similarity score. We can see that the viruses from India and China are estimated to have, quite literally, an estimate of 100% similarity as well as the intersection between these two and the French virus. India and China are very close regions, which would make sense. France being so similar, however, can leave us somewhat baffled as to what this implies.

Furthermore, we can see high similarity scores bewteen the United Kingdom and the intersection with Italy, Germany and South Korea. This, coupled with the previous statements, suggests that there was a lot of immigration from the east to the west or vice-versa during the times in which these viruses were active.

Finally, comparing the mortality rates, we can see that the most violent virus strands were located in asiatic countries, that is, South Korea, China, India and Japan. There were some very violent viruses in Europe, too, but their density is not as high as in Asia. America, the continent, seems to have had relatively mild mortality rates, although of course they were still high.